feat: chain source rest sync with bitcoind

enigbe opened the pull request.

enigbe commented:
Thanks for the early feedback.

> 1. I'm not quite convinced we should rename the BitcoindRpc client. Wouldn't it be cleaner to just add a (granted, rather duplicative) BitcoindRest variant side-by-side, and then in a second step maybe see how we can DRY up the code. Mushing both variants together seems to result in a rather confusing API and code structure.
> 2. I'm a bit confused why we'd now still have RPC credentials on set_chain_source_bitcoind (and elsewhere in downstream from that), even if we want to use the REST backend? Is there any operation that still requires RPC access? Can't we just rely on REST alone?

I understand your concerns about the name and API changes I have introduced but here is a bit of context.

Adding a `ChainSource::BitcondRest` variant was my initial approach, too because I had erroneously thought Bitcoin Core's [REST interface](https://github.com/bitcoin/bitcoin/blob/master/doc/REST-interface.md) exposed (more or less) the same APIs as the RPC interface. However, the REST interface exposes a handful of APIs only for `GET` requests. It is not possible to [broadcast transactions](https://github.com/lightningdevkit/ldk-node/blob/1525255222cd7a48de3a9301a3053e8a4e3d1a86/src/chain/bitcoind_rpc.rs#L56-L60) with a pure REST variant as we do with the `BitcoindRpcClient`, There is no REST API to `sendrawtransaction`.
For that particular use case, given that we'd need a handle on an RPC client regardless, I thought it best to:
- Keep the `BitcoindRpcClient` as is.
- Introduce a configurable REST/RPC sync client. With this, users can decide if syncing should happen either via the REST or RPC APIs.
- Rename `ChainSource::BitcoindRpc` to `ChainSource::Bitcoind` as the latter leaves no ambiguity about the possibility of a REST variant.

Camillarhi commented:
Hi @enigbe,

I noticed your CI is failing. I had a similar issue, but it was resolved after rebasing onto the latest `dev` branch (which includes changes from https://github.com/lightningdevkit/ldk-node/pull/524).

Could you try rebasing and see if that fixes the problem?

Thanks

chuksys commented on src/builder.rs:
```diff
@@ -297,16 +310,70 @@ impl NodeBuilder {
+					|| rpc_user.is_empty()
+					|| rpc_password.is_empty()
+				{
+					panic!("RPC configuration is incomplete. Must fully configure via RPC first.");
```
I think we could introduce a new `BuildError` variant like `IncompleteRPCConfig` and error out instead?

chuksys commented on src/builder.rs:
```diff
@@ -297,16 +310,70 @@ impl NodeBuilder {
+				*sync_client_config = Some(rest_client_config);
+			},
+			Some(cs) => {
+				panic!("This option is only valid for ChainSource::Bitcoind. Current chain source config: {cs:?}")
```
Maybe we could introduce a new `BuildError` variant like `InvalidChainSource` and error out instead?

chuksys commented on src/builder.rs:
```diff
@@ -297,16 +310,70 @@ impl NodeBuilder {
+				panic!("This option is only valid for ChainSource::Bitcoind. Current chain source config: {cs:?}")
+			},
+			None => {
+				panic!("Must configure via RPC first.")
```
Here too I think we could error out by introducing and returning a new `BuildError` variant like `RPCNotConfigured`.

chuksys reviewed (commented):
Looking into this PR has helped me gain a good understanding of how chain sourcing works. I think introducing `BitcoindSyncClient` is a good idea as the user can still carry out syncing operations using the REST API while the RPC Client is still maintained to carry out operations not supported by the REST API. I have just a few suggestions with regards to error handling.

enigbe commented:
Thanks for the feedback @tnull. I revisited this last week and addressed concerns raised here.
> Aaah, okay, then the proposed approach makes more sense. I think it's still a bit awkward to have the mandatory RPC credentials on the `set_chain_source_bitcoind` method and _addtionally_ have the `BitcoindSyncConfig::Rpc` variant.
> 
This has been removed and we just keep a [config for REST sync](https://github.com/lightningdevkit/ldk-node/blob/a28d28fca6d594c8c0c7aaff30c4caaad8c53d5d/src/config.rs#L402-L407).

> While it then makes sense to reuse the client internally, it might be simpler to just keep the original `set_chain_source_bitcoind_rpc` method and just add a `set_chain_source_bitcoind_rest` variant, that also features the mandatory RPC credentials?

Agreed! This seems like a cleaner split. I've added a `set_chain_source_bitcoind_rest` variant accordingly. However, I think it'd be potentially confusing/awkward for callers to still pass RPC parameters to the REST function.  I opted for `set_chain_source_bitcoind_rest` to take only REST parameters and clarified comments on usage and dependency on the RPC variant.

If you'd still prefer we pass RPC parameters as a required input for the REST variant, happy to adjust.

enigbe commented:
> Hi @enigbe,
> 
> I noticed your CI is failing. I had a similar issue, but it was resolved after rebasing onto the latest `dev` branch (which includes changes from #524).
> 
> Could you try rebasing and see if that fixes the problem?
> 
> Thanks

Thanks for looking at this. I've rebased and have just the CLN integration tests failing. I tried reproducing the failing test locally but couldn't. I should give it another go today.

enigbe commented:
> Looking into this PR has helped me gain a good understanding of how chain sourcing works. I think introducing `BitcoindSyncClient` is a good idea as the user can still carry out syncing operations using the REST API while the RPC Client is still maintained to carry out operations not supported by the REST API. I have just a few suggestions with regards to error handling.

Thanks for the review. Added the suggestions recommended for error handling.

enigbe commented on src/builder.rs:
```diff
@@ -297,16 +327,72 @@ impl NodeBuilder {
+	///
+	/// ## Parameters:
+	/// * `rest_host`, `rest_port` - Required parameters for the Bitcoin Core REST connection.
+	pub fn set_chain_source_bitcoind_rest(
```
I'll refactor to include the recommended changes . I thought having to pass RPC config for REST would be confusing but I now agree that it would be clearer and less likely to misuse.

enigbe commented:
Please excuse the delay on this. I have addressed concerns from the last review. Investigating a flaky test but it's ready for another look.

Let me know what you think when you can review.

enigbe commented on src/chain/bitcoind/mod.rs:
```diff
@@ -0,0 +1,533 @@
+
+pub mod common;
+
+pub enum BitcoindApiClient {
```
This has been renamed.

enigbe commented on src/chain/bitcoind.rs:
```diff
@@ -272,35 +514,100 @@ impl BitcoindRpcClient {
-		Box::pin(async move { self.rpc_client.get_best_block().await })
+	fn get_best_block(
+		&self,
+	) -> lightning_block_sync::AsyncBlockSourceResult<(bitcoin::BlockHash, Option<u32>)> {
```
This has been addressed. I inadvertently qualified the type instead of importing it.

enigbe commented:
Hi @tnull,

This is ready for another look. As recommended, I have added:
- prefector commit where the previously introduced macros are refactored into helper methods in preparation for REST support.
- couple of fix-ups to address unwanted code move and type imports.

enigbe commented on src/chain/bitcoind.rs:
```diff
@@ -272,35 +514,100 @@ impl BitcoindRpcClient {
 }
 
+#[derive(Debug, Clone)]
+pub(crate) struct MempoolEntry {
```
Possibly moved this while rebasing but it's now addressed.
